cpp/merkletree/sparse_merkle_tree.h

```diff
@@ -0,0 +1,265 @@
+  virtual std::string LeafHash(const std::string& data) const {
+    return treehasher_.HashLeaf(data);
+  }
+
+  // Add a new leaf to the hash tree. Stores the hash of the leaf data in the
+  // tree structure at path |Hash(key)|, does not store the data itself.
+  //
+  // @param data Binary input blob
+  // @param path Binary path of node to set.
+  virtual void SetLeafByHash(const std::string& key, const std::string& data);
+
+  // Add a new leaf to the hash tree. Stores the hash of the leaf data in the
+  // tree structure, does not store the data itself.
+  //
+  // @param data Binary input blob
+  // @param path Binary path of node to set.
+  virtual void SetLeaf(const Path& path, const std::string& data);
+
+  // Get the current root of the tree.
+  // Update the root to reflect the current shape of the tree,
+  // and return the tree digest.
+  //
+  // Returns the hash of an empty string if the tree has no leaves
+  // (and hence, no root).
+  virtual std::string CurrentRoot();
+
+  // Get the Merkle path from the leaf at |path| to the current root.
+  //
+  // Returns a vector of node hashes, ordered by levels from leaf to root.
+  // The first element is the sibling of the leaf hash, and the last element
+  // is one below the root.
+  // Returns an empty vector if the tree is not large enough
+  // or the leaf index is 0.
+  //
+  // @param path the path of the leaf whose inclusion proof to return.
+  std::vector<std::string> InclusionProof(const Path& path);
+
+  std::string Dump() const;
+
+ private:
+  // WARNING WARNING WARNING
+  // 64 < 256 !
+  // WARNING WARNING WARNING
+  // TODO(alcutter): BIGNUM probably.
+  typedef uint64_t IndexType;
+
+  struct Leaf {
+    Leaf(const Path& path, const std::string& value)
+        : path_(path), value_(value) {
+    }
+
+    std::string DebugString() const;
+
+    const Path path_;
+    std::string value_;
+  };
+
+  struct TreeNode {
+    TreeNode() : type_(INTERNAL), hash_(nullptr) {
+    }
+    TreeNode(const std::string& hash)
+        : type_(INTERNAL), hash_(new std::string(hash)) {
+    }
+
+    TreeNode(const Path& path, const std::string& leaf_value)
+        : type_(LEAF), leaf_(new Leaf(path, leaf_value)) {
+    }
+
+    std::string DebugString() const;
+
+    enum { INTERNAL, LEAF } type_;
+    // TODO(alcutter): sort this out
+    std::unique_ptr<std::string> hash_;
+    std::unique_ptr<Leaf> leaf_;
+  };
+
+  std::string CalculateSubtreeHash(size_t depth, IndexType index);
+
+  void DumpTree(std::ostream* os, size_t depth, IndexType index) const;
+
+  // Get the |index|-th node at level |level|. Indexing starts at 0;
+  // caller is responsible for ensuring tree is sufficiently up to date.
+  std::string Node(size_t level, size_t index) const;
+
+  // Maybe add a new tree level.
+  void EnsureHaveLevel(size_t n);
+
+  std::unique_ptr<SerialHasher> serial_hasher_;
+  TreeHasher treehasher_;
+  const std::vector<std::string>& null_hashes_;
```

Comment:
I find this line very disturbing to look at, because const references are allowed to be temporaries, but putting them in a `struct`/`class` doesn't extend the lifetime of the temporaries. I think.

http://en.cppreference.com/w/cpp/language/reference_initialization#Lifetime_of_a_temporary

Might be best to be dealing in (const) pointers for this?